### DISCUSSION TITLE
How can I make multilanguage bot using laravel localization

### DISCUSSION THREAD

**Discussion by foxdevuz** [nutgram/nutgram#789] [Q&A]
I am using laravel localization to create multilanguage bot, and it's not working for the language that shown in `.env` file here's the code:

First I am checking what's the users prefered language from database using middleware
```
class GetLanguage
{
    public function __invoke(Nutgram $bot, $next): void
    {
        $user_lang = User::where("user_id", $bot->userId())->first()->lang ?? 'en';
        App::setLocale($user_lang);

        $next($bot);
    }
}
```

after that, checking it using `$bot->onText()` function here's one of them:
```
$bot->onText(__("admin.keyboard.del_admin"), DeleteAdminConversation::class)->middleware(CheckAdmin::class)->middleware(GetLanguage::class);
```

however I can't get a response for the languages except shown in `.env` file.

what should I do and how can I get the result that I want?)

**Comment by Lukasss93** [comment]
Nutgram routes are resolved at application startup.

Just do this:
```php
class LaravelTrans
{
    public function __construct(protected string $key) {
    }

    public function __invoke(Nutgram $bot, $next)
    {
        if($bot->message()?->text !== trans($this->key)){
            return;
        }
        
        $next($bot);
    }
}
```
```php
$bot->onMessage(KeyboardButtonA::class)->middleware(new LaravelTrans('button.a'));
$bot->onMessage(KeyboardButtonB::class)->middleware(new LaravelTrans('button.b'));
$bot->onMessage(KeyboardButtonC::class)->middleware(new LaravelTrans('button.c'));
```

![348830305-cf6d21e0-6be5-4f02-8faf-84105609c29a](https://github.com/user-attachments/assets/0c052c2e-319f-4b2b-9884-097665518081)

**Reply by foxdevuz** [reply]
oh, I see. 
I understand how it works now. but what if I want t make a conversation? not only handler... 

for example if I am adding FAQ I should ask for qustion and answer. so how can I work in this case?

**Reply by Lukasss93** [reply]
```php
class FAQ extends Conversation
{
    public function start(Nutgram $bot): void
    {
        $bot->sendMessage(
            text: 'Enter a question from the keyboard:',
            reply_markup: ReplyKeyboardMarkup::make()
                ->addRow(KeyboardButton::make(trans('question.foo')))
                ->addRow(KeyboardButton::make(trans('question.bar')))
                ->addRow(KeyboardButton::make(trans('close.keyboard')))
        );

        $this->setSkipHandlers(true);
        $this->next('waitAction');
    }

    public function waitAction(Nutgram $bot): void
    {
        match($bot->message()?->text) {
            trans('question.foo') => $this->answerFoo($bot),
            trans('question.bar') => $this->answerBar($bot),
            trans('close.keyboard') => $this->closeFaq($bot),
            default => $this->start($bot),
        };
    }

    public function answerFoo(Nutgram $bot): void
    {
        $bot->sendMessage(trans('answer.foo'));
    }

    public function answerBar(Nutgram $bot): void
    {
        $bot->sendMessage(trans('answer.bar'));
    }

    public function closeFaq(Nutgram $bot): void
    {
        $bot->sendMessage(
            text: 'Thanks for using the FAQ!',
            reply_markup: ReplyKeyboardRemove::make(true),
        );

        $this->end();
    }
}

``` 

https://github.com/user-attachments/assets/8f45f134-78e9-40be-bb42-d5f770f69cad

**Reply by foxdevuz** [reply]
actually I meant how to do it using multilanguage, now only with chat. I did it with this way.
- I handled the message in handler and redirected it to necessary conversation like this way
```
$bot->onMessage(AddAdminHandler::class)
    ->middleware(CheckAdmin::class)
    ->middleware(new LaravelTrans("admin.keyboard.add_admin"));
 ```
 
 ```
 class AddAdminHandler
{
    /**
     * @throws InvalidArgumentException
     */
    public function __invoke(Nutgram $bot): void
    {
        (new AddAdminConversation())->start($bot);
    }
}
```

and it's working, but there's one problem, I am getting this Error on console:
![image](https://github.com/user-attachments/assets/cedffd7a-9221-4bbb-ad78-27e8f263bc1e)

What am I supposed to do? is there any better way?

**Reply by Lukasss93** [reply]
To start a conversion, just enter this:
```php
$bot->onCommand('admin', AddAdminConversation::class);
```
OR
```php
$bot->onMessage(AddAdminConversation::class);
```
OR
```php
AddAdminConversation::begin($bot);
```

Please read our documentation website:
https://nutgram.dev/docs/usage/conversations#begin-a-conversation

**Reply by foxdevuz** [reply]
Wait I will explain it what I mean:
if I use conversation with the way you said for localization, it's not going out from conversation here's one example
telegram.php
```
$bot->onMessage(AddAdminConversation::class)
    ->middleware(CheckAdmin::class)
    ->middleware(new LaravelTrans("admin.keyboard.add_admin"));

$bot->onMessage(DeleteAdminConversation::class)
    ->middleware(CheckAdmin::class)
    ->middleware(new LaravelTrans("admin.keyboard.del_admin"));
```

LaravelTrans:
```
class LaravelTrans
{
    public function __construct(protected string $key) {
    }

    public function __invoke(Nutgram $bot, $next): void
    {
        $user_lang = User::where("user_id", $bot->userId())?->first()?->lang ?? 'en';
        app()->setLocale($user_lang);

        if($bot->message()?->text !== trans($this->key)){
            return;
        }

        $next($bot);
    }
}
```
DelAdminConversation:
```
class DeleteAdminConversation extends Conversation
{
    use DeleteAdmin;

    /**
     * @throws InvalidArgumentException
     */
    public function start(Nutgram $bot): void
    {
        $bot->sendMessage(
            text: $this->list_admins(),
            parse_mode: ParseMode::MARKDOWN_LEGACY
        );
        $this->next('delAdmin');
    }
*** // rest of the code
```
Add Admin Conversation
```
class AddAdminConversation extends Conversation
{
    /**
     * @throws InvalidArgumentException
     */
    public function start(Nutgram $bot): void
    {
        $bot->sendMessage(
            text: __("add_admin.send_admin_name")
        );
        $this->next('getAdminName');
    }
    *** // rest of the code
```

and when I click `Add Admin` button it's working as expected, cool, but if I click `Del Admin` button before adding new admin (it's like I am in the AddAdminConversation and without finishing it I am gonna delete admin), it's not going out from AddAdminConversation, and it's finding the del admin button as the user response and going next action in AddAdminConversation.

Here's the image:
![image](https://github.com/user-attachments/assets/10bbe64d-00cf-44f1-a4f8-7638524a74d5)
 
 it's finding the `⛔️ Delete admin` as the admin's name not the button and not handling it. 
 
 There's the explanation in graph(I am not too good at drawing graphs)
![image](https://github.com/user-attachments/assets/779bb492-f270-4e35-9c52-9fd2ef4edb2e)

I have some solutions but I don't think it is best practice:
- on the second function of each Conversation I am checking if the text exists in buttons array and if so, I am ending conversation
- handle the button in handler and redirect it to conversation (as I send it above)


I hope I explained what I meant...

**Reply by Lukasss93** [reply]
You can't do that.
If you are inside a conversation, you can't exit without doing it explicitly through `$this->end()` 
or through a command with `$bot->endConversation()` inside.

You should handle everything in one conversation.

Look at this [example](https://github.com/Lukasss93/telegram-stickeroptimizer/blob/master/app/Telegram/Conversations/SettingsConversation.php), it is an InlineMenu but the concept is the same.
You can test it by running the /setting command on https://t.me/NewStickerOptimizerBot

**Reply by foxdevuz** [reply]
Okay, thanks for the answer and help!

**Reply by foxdevuz** [reply]
wait, I have found another way to it and it kinda helped but still have some problems here's how I did:
`TranslateMiddleware:`
```
class TranslateMiddleware
{
    use CheckIFButtonIsOntheUpdate;

    /**
     * @throws InvalidArgumentException
     */
    public function __invoke(Nutgram $bot, $next): void
    {
        $user_lang = User::where("user_id", $bot->userId())?->first()?->lang ?? 'en';
        app()->setLocale($user_lang);

        if($this->isUserInConversationAndEndIfTextMatchesButtons($bot?->message()?->text)) {
            $bot->endConversation($bot->chatId());
            $bot->sendMessage("User is not in converation");
        } else {
            $bot->sendMessage("User in conversation");
        }
        $next($bot);
    }
}
```


`CheckIFButtonIsOntheUpdate:`
```
trait CheckIFButtonIsOntheUpdate
{
    protected function isUserInConversationAndEndIfTextMatchesButtons(string $text): bool
    {
        $buttonOptions = [
            __("keyboards.user_main.support"),
            __("keyboards.user_main.faq"),
            __("admin.keyboard.faq"),
            __("admin.keyboard.faq_list"),
            __("admin.keyboard.add_admin"),
            __("admin.keyboard.del_admin"),
        ];

        return in_array($text, $buttonOptions);
    }
}
```

and I am using this middleware as global middleware
`$bot->middleware(TranslateMiddleware::class);`

and some examples for routing
```
$bot->onMessage(AddFaqConversation::class)
    ->middleware(CheckAdmin::class)
    ->middleware(new LaravelTrans("admin.keyboard.faq"));

$bot->onMessage(FAQListConversation::class)
    ->middleware(CheckAdmin::class)
    ->middleware(new LaravelTrans("admin.keyboard.faq_list"));

$bot->onMessage(AddAdminConversation::class)
    ->middleware(CheckAdmin::class)
    ->middleware(new LaravelTrans("admin.keyboard.add_admin"));

$bot->onMessage(DeleteAdminConversation::class)
    ->middleware(CheckAdmin::class)
    ->middleware(new LaravelTrans("admin.keyboard.del_admin"));
```

The idea:

as written in docs https://nutgram.dev/docs/usage/middleware the global middleware should work before anything, in my case I am checking if the text is in the buttons array and if so, then user clicked the button and it should end the conversation and start the new one. it's working as expected which means it's ending the conversation when another button clicks but... there's one problem.

![image](https://github.com/user-attachments/assets/a7e2fa6b-c8a3-484f-bdc9-9fee12cd8038)

The `User in conversation` message has been sent when I choose the language, it should have continues the conversation but somehow it's ending as soon as it's opened (I see the database and there's nothing in cache).

The problem is here, it's working half way, kinda it's working and ending conversation if the text is in the buttons array but it's also ending the conversation as soon as it started.

NOTE: in the image you can see it's sending several message to me, it's not because I used it in several places in my code I used in only one file, it's `telegram.php`


here's the conversation to add FAQ:
```
class AddFaqConversation extends Conversation
{
    use SelectLanguage, AdminKeyboards, AdminHelpers;

    /**
     * @throws InvalidArgumentException
     */
    public function start(Nutgram $bot): void
    {
       $this->chose_lang($bot);
    }

private function chose_lang(Nutgram $bot, bool $repeating = false): void
    {
        $bot->sendMessage(
            text: $repeating ? trans("faq.start") : trans("faq.choose_lang"),
            parse_mode: ParseMode::MARKDOWN_LEGACY,
            reply_markup: $this->chooseLanguage()
        );
        $this->next('getLangForFAQ');
    }

// rest of the code
```

as you can see I am not ending it on the conversation.


How can this happen and what's the solution?

ps: yes your example to do it all in one conversation with Inlin Keyboard is cool but I need to rewrite everything... I am kinda lazy ngl xD

What I Mean as graph:

![image](https://github.com/user-attachments/assets/8b43605b-c71f-4771-a4c9-7566ace43b2e)
